Application/ControlMsrE2/ControlMsrE2.h

```diff
@@ -0,0 +1,230 @@
+///
+
+#include <Uefi.h>
+#include <PiDxe.h>
+#include <Library/PcdLib.h>
+#include <Library/UefiLib.h>
+#include <Library/BaseMemoryLib.h>
+#include <Library/DebugLib.h>
+#include <Library/OcMiscLib.h>
+#include <Library/UefiBootServicesTableLib.h>
+#include <Library/UefiRuntimeServicesTableLib.h>
+#include <Library/UefiApplicationEntryPoint.h>
+
+#include <Library/MemoryAllocationLib.h>
+#include <Uefi/UefiInternalFormRepresentation.h>
+#include <Library/UefiHiiServicesLib.h>
+#include <Library/HiiLib.h>
+
+#include <Protocol/HiiFont.h>
+#include <Protocol/HiiImage.h>
+#include <Protocol/HiiImageEx.h>
+#include <Protocol/HiiImageDecoder.h>
+#include <Protocol/HiiString.h>
+#include <Protocol/HiiDatabase.h>
+#include <Protocol/HiiConfigRouting.h>
+#include <Protocol/HiiConfigAccess.h>
+#include <Protocol/HiiConfigKeyword.h>
+#include <Guid/HiiPlatformSetupFormset.h>
+
+#define DONT_STOP_AT 0xFFFF
+
+/**
+  Parameters to search in a Form for
+  IfrHeaders which contain a Description
+  and to return the VarStore
+
+  @param SearchText        Text that must be part of the options
+                           description
+  @param EfiHandle         Handle of the list the Form is part of
+  @param ListHeader        Pointer to the contents of the handle
+  @param PkgHeader         Pointer to the current package in the
+                           list
+  @param FirstIfrHeader    Pointer to first IfrHeader in Package
+  @param IfrVarStore       Pointer to IfrHeader of corresponding
+                           varstore in same list
+  @param IfrOneOf          Pointer to IfrHeader of BIOS Option
+  @param StopAt            Find a certain BIOS Option
+  @param Count             Running number of suitable BIOS Options
+**/
+typedef struct ONE_OF_CONTEXT_ {
+  EFI_STRING                    SearchText;
+  EFI_HII_HANDLE                EfiHandle;
+  EFI_HII_PACKAGE_LIST_HEADER   *ListHeader;
+  EFI_HII_PACKAGE_HEADER        *PkgHeader;
+  EFI_IFR_OP_HEADER             *FirstIfrHeader;
+  EFI_IFR_VARSTORE              *IfrVarStore;
+  EFI_IFR_ONE_OF                *IfrOneOf;
+  UINT16                        StopAt;
+  UINT16                        Count;
+} ONE_OF_CONTEXT;
+
+/**
+  Callback - What to do, when a IfrHeader with a defined
+  OP_CODE is found
+
+  @param[in]     IfrHeader   Current IfrHeader under scrutiny
+  @param[in,out] Stop        Pointer to Stop flag. If TRUE no
+                             further search for opcodes. Can
+                             be NULL.
+  @param[in]     Context     Pointer to Handler specific data
+**/
+typedef
+VOID
+OP_CODE_HANDLER (
+  IN     EFI_IFR_OP_HEADER *IfrHeader,
+  IN OUT BOOLEAN           *Stop  OPTIONAL,
+  IN     VOID              *Context
+  );
+
+/**
+  Commandline Arguments
+**/
+enum {
+  ARG_VERIFY      = 0,
+  ARG_LOCK        = 1,
+  ARG_UNLOCK      = 2,
+  ARG_INTERACTIVE = 8,
+};
+
+#define PADD(x,y) (VOID *)(((CHAR8 *) x) + y)
```

Comment:
In my opinion we either need something like this globally, or not at all. Should discuss. cc @vit9696